### PR TITLE
fix(gpu): don't dispose a forward intermediate aliased by the next layer's reshape view

### PR DESCRIPTION
**Pull request by charlesburford**
## What & why

On the OpenCL `DirectGpuTensorEngine`, `Predict` (GPU-resident inference via `ForwardGpu`)
returns an **all-zero** output for any network containing a layer that returns a reshape
**view** — most commonly `FlattenLayer`. Because the failure is silent and lives only in the
inference forward, a **correctly-trained GPU model reads as "learning nothing"** (accuracy parks
at chance, loss frozen). The same model on `CpuEngine` is correct. Training itself is unaffected
— only the GPU inference path is.

## Root cause

`NeuralNetworkBase.ForwardGpu` disposed each intermediate immediately after the next layer
consumed it:

```csharp
var next = layer.ForwardGpu(current);
if (ownsCurrentTensor && current is not null) current.Dispose();
current = next;
```

`FlattenLayer.ForwardGpu` returns `current.Reshape(...)` — a **zero-copy view** that shares
`current`'s backing `Vector`, GPU buffer, and deferred-download materializer. Disposing `current`
frees that buffer and removes the materializer the view still needs, so the next layer reads a
freshly-materialized **zero** array → the following `Dense` GEMM computes `0·W + bias` →
all-zeros. It reproduces on OpenCL specifically because OpenCL doesn't take the CUDA-only
resident-into fast path and flows through this generic chain. Same bug class as the existing
`#1625` fix a few lines above.

## Fix

Guard the in-loop dispose: when `next` aliases `current`'s backing (`ReferenceEquals` on the
shared `Vector`), rebind `next` onto an **independent** `Tensor<T>.FromGpuBuffer` over the same
buffer and transfer ownership before disposing `current` — so the dispose neither strands `next`
nor double-frees. The guard is true only for the zero-copy view path, so normal GPU layers
(Dense/Conv, which return fresh `FromGpuBuffer` tensors) are unaffected.

```csharp
var next = layer.ForwardGpu(current);

// Dispose intermediate if we own it (but not the input).
if (ownsCurrentTensor && current is not null)
{
    // If `next` is a zero-copy view sharing current's backing (e.g.
    // FlattenLayer's Reshape), disposing current frees the buffer the
    // view still needs -> the next layer reads all-zeros. Rebind `next`
    // onto an independent tensor over the same buffer and hand off
    // ownership so Dispose() is safe (no strand, no double-free).
    if (ReferenceEquals(next.DataVector, current.DataVector)
        && current._gpuBuffer is not null
        && current._gpuBackend is not null)
    {
        var rebound = Tensor<T>.FromGpuBuffer(
            current._gpuBackend, current._gpuBuffer, next.Shape.ToArray(),
            GpuTensorRole.Activation, ownsBuffer: current._ownsGpuBuffer);
        current._ownsGpuBuffer = false;
        next = rebound;
    }

    current.Dispose();
}

current = next;
ownsCurrentTensor = true;
```

## Proof — before/after

Controlled A/B: same RTX 2060, same `Active engine : DirectGpuTensorEngine`, same code/data/seed.
The only variable changed is this patch.

**Untrained `Flatten→Dense` forward — raw logits (no training):**

| | logits |
|---|---|
| **Before** (GPU, unpatched) | `[0, 0, 0, 0]` — input-independent zeros |
| **After** (GPU, patched) | `[-0.106698, 0.584231, -0.296298, -0.0384265]` |
| CPU reference | `[-0.106698, 0.584231, -0.296298, -0.0384265]` |

After the fix the GPU forward is **bit-identical to the CPU**.

**Training accuracy, 4 architectures, 6 epochs (chance = 25%):**

| architecture | Before (GPU) | After (GPU) |
|---|---|---|
| `Flatten→Dense` | 25% | **100%** |
| `1×Conv→Dense` | 25% | **100%** |
| `3×Conv (no pool)` | 25% | **100%** |
| `3×(Conv+MaxPool)` (full CNN) | 25% | **100%** |

Control: on `CpuEngine` every rung is 100% before *and* after — so the 25%→100% jump is solely
the fixed GPU path.

## Minimal repro

```csharp
var layers = new List<ILayer<float>>
{
    new FlattenLayer<float>(),
    new DenseLayer<float>(outputSize: 4, activationFunction: new SoftmaxActivation<float>()),
};
var arch = new NeuralNetworkArchitecture<float>(
    inputType: InputType.ThreeDimensional,
    taskType: NeuralNetworkTaskType.MultiClassClassification,
    complexity: NetworkComplexity.Simple,
    inputHeight: 16, inputWidth: 16, inputDepth: 1, outputSize: 4, layers: layers);
var net = new ConvolutionalNeuralNetwork<float>(arch, new CategoricalCrossEntropyLoss<float>());

var x = new Tensor<float>(new[] { 1, 1, 16, 16 });
for (int i = 0; i < x.Length; i++) x[i] = 0.5f;
var y = net.Predict(x);   // GPU: 0,0,0,0  | CPU: a valid softmax
Console.WriteLine(string.Join(", ", Enumerable.Range(0, 4).Select(k => y[0, k])));
```

(Requires a real OpenCL device — `DirectGpuTensorEngine` auto-selects on native Windows. WSL has
no OpenCL ICD and falls back to CPU. `AIDOTNET_DISABLE_GPU=1` forces the CPU control.)

## Scope / caveat

This fixes the "GPU learns nothing" bug — classification (argmax) is now 100% on GPU, matching
the CPU. It **revealed** a separate, pre-existing GPU **softmax** bug (the activation returns
non-normalized/negative probability values; does **not** affect argmax/accuracy), which was
masked while the forward returned zeros. That is independent of this change and will be filed
separately.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved GPU forward-pass handling to avoid accidentally disposing intermediate tensors when consecutive layers share the same underlying GPU memory.
  * Fixed an issue where zero-copy tensor views could be invalidated during execution, improving output stability on GPU.

### CODE REVIEW

File: src/NeuralNetworks/NeuralNetworkBase.cs [hunk 1]
```diff
@@ -1320,9 +1320,25 @@ public virtual Tensor<T> ForwardGpu(Tensor<T> input)
...
+                            && current._gpuBuffer is not null
+                            && current._gpuBackend is not null)
+                        {
+                            var rebound = Tensor<T>.FromGpuBuffer(
+                                current._gpuBackend, current._gpuBuffer, next.Shape.ToArray(),
+                                GpuTensorRole.Activation, ownsBuffer: current._ownsGpuBuffer);
+                            current._ownsGpuBuffer = false;
+                            next = rebound;
+                        }
+
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/NeuralNetworks/NeuralNetworkBase.cs]: Addressed in 8db1e9d5a: extracted the rebind into a shared `RebindGpuAliasBeforeDispose` helper and applied it to `ForwardDeferred` and `ForwardDeferredAsync` as well, which had the same dispose-after-consume pattern. `ForwardWithGpuContext` is intentionally left unchanged — it doesn't dispose `current` in-loop (the GpuExecutionContext registry owns disposal), so there's no stranded-view window there. Builds net10.0/net8.0/net471.

**Review by ooples — APPROVED** [review]
